### PR TITLE
New ambient occlusion effect in render pipeline

### PR DESCRIPTION
**Pull request by venkatn93**
Removing old naked GL implementation of screen space AO and replacing it with an SSAO (soon to be HBAO) implementation that is part of the graphics pipeline. This AO effect is still a work in progress, and can be enabled/disabled from Developer -> Render -> Debug Ambient Occlusion

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 075c9f05de9a3b340554f1792c35a74288ca7fcd available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 1]
```diff
@@ -0,0 +1,404 @@
+//
+//  AmbientOcclusionEffect.cpp
+//  interface/src/renderer
+//
+//  Created by Andrzej Kapolka on 7/14/13.
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: change this to be your name and new date.

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+//  Copyright 2013 High Fidelity, Inc.
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: update this

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+#include "gaussian_blur_vertical_vert.h"
+#include "gaussian_blur_horizontal_vert.h"
+#include "gaussian_blur_frag.h"
+#include "occlusion_blend_vert.h"
+#include "occlusion_blend_frag.h"
+
+const int ROTATION_WIDTH = 4;
+const int ROTATION_HEIGHT = 4;
+
+/*
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: remove dead code

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        auto ps = gpu::ShaderPointer(gpu::Shader::createPixel(std::string(ambient_occlusion_frag)));
+        gpu::ShaderPointer program = gpu::ShaderPointer(gpu::Shader::createProgram(vs, ps));
+
+        gpu::Shader::BindingSet slotBindings;
+        slotBindings.insert(gpu::Shader::Binding(std::string("depthTexture"), 0));
+        slotBindings.insert(gpu::Shader::Binding(std::string("normalTexture"), 1));
+
+        gpu::Shader::makeProgram(*program, slotBindings);
+
+        //_drawItemBoundPosLoc = program->getUniforms().findLocation("inBoundPos");
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: dead code?

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+    glm::vec2 bottomLeftSmall(0.5f, -1.0f);
+    glm::vec2 topRightSmall(1.0f, -0.5f);
+    DependencyManager::get<GeometryCache>()->renderQuad(batch, bottomLeftSmall, topRightSmall, texCoordTopLeft, texCoordBottomRight, color);
+    
+    // Ready to render
+    args->_context->syncCache();
+    renderContext->args->_context->syncCache();
+    args->_context->render((batch));
+
+    // need to fetch forom the z buffer and render something in a new render target a result that combine the z and produce a fake AO result
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: what is this comment about? also couple of typos here.

File: libraries/render-utils/src/AmbientOcclusionEffect.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
+//
+//  AmbientOcclusionEffect.h
+//  interface/src/renderer
+//
+//  Created by Andrzej Kapolka on 7/14/13.
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.h]: update

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 7]
```diff
@@ -59,6 +60,7 @@ RenderDeferredTask::RenderDeferredTask() : Task() {
     _jobs.push_back(Job(new ResetGLState::JobModel()));
     _jobs.push_back(Job(new RenderDeferred::JobModel("RenderDeferred")));
     _jobs.push_back(Job(new ResolveDeferred::JobModel("ResolveDeferred")));
+    _jobs.push_back(Job(new AmbientOcclusion::JobModel("AmbientOcclusion")));
```
Comment (ZappoMan) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: do we only want to do this if the menu item is enabled?

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 8]
```diff
@@ -72,10 +74,14 @@ RenderDeferredTask::RenderDeferredTask() : Task() {
     _jobs.push_back(Job(new DrawTransparentDeferred::JobModel("TransparentDeferred", _jobs.back().getOutput())));
     
     _jobs.push_back(Job(new render::DrawStatus::JobModel("DrawStatus", renderedOpaques)));
+
+
     _jobs.back().setEnabled(false);
     _drawStatusJobIndex = _jobs.size() - 1;
 
+    //_jobs.push_back(Job(new AmbientOcclusion::JobModel("AmbientOcclusion")));
```
Comment (ZappoMan) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: what's this?

File: libraries/render-utils/src/ambient_occlusion.slf [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+varying vec2 varTexcoord;
+
+uniform sampler2D depthTexture;
+uniform sampler2D normalTexture;
+
+float getRandom(vec2 co) {
+    return fract(sin(dot(co.xy ,vec2(12.9898,78.233))) * 43758.5453);
+}
+/*
```
Comment (ZappoMan) [inline libraries/render-utils/src/ambient_occlusion.slf]: dead code?

File: libraries/render-utils/src/ambient_occlusion.slf [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+void main(void) {
+
+    vec4 depthColor = texture2D(depthTexture, varTexcoord.xy);
+	vec4 normalColor = texture2D(normalTexture, varTexcoord.xy);
+    float z = depthColor.r;                       // fetch the z-value from our depth texture
+    float n = 1.0;                                // the near plane
+    float f = 30.0;                               // the far plane
+    float c = (2.0 * n) / (f + n - z * (f - n));  // convert to linear values 
+
+    //gl_FragColor = vec4(c, c, c, 1.0);
```
Comment (ZappoMan) [inline libraries/render-utils/src/ambient_occlusion.slf]: dead code?

File: libraries/render-utils/src/gaussian_blur.slf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+// the interpolated normal
+//varying vec4 interpolatedNormal;
+
+varying vec2 varTexcoord;
+varying vec2 varBlurTexcoords[14];
+
+uniform sampler2D occlusionTexture;
+
+void main(void) {
+	gl_FragColor = vec4(0.0);
```
Comment (ZappoMan) [inline libraries/render-utils/src/gaussian_blur.slf]: indent weird

**Comment by hifi-gustavo** [comment]
Build console output for commit: e0634de4032c7fce30606cf411b5f35ebe935b7f available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9ccf316513ee580c44976ee50f8665b07bd0fa34 available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 216c499d1472abecc29968bd9f2ba6b0aa7be7ac available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e0366027269654694b706e4896044d5006dacbe0 available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fc612ab8cd03997dc767df3e4e72a0240ab22ffe available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5dd16d9f801cb79cbaa786d1582a3eeca16a443a available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6bc3d65bf4cfe58e3ba12a75979baa13ce3228ce available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

File: libraries/gpu/src/gpu/GLBackendTexture.cpp [hunk 12]
```diff
@@ -144,6 +144,38 @@ class GLTexelFormat {
                 case gpu::RGB:
                 case gpu::RGBA:
                     texel.internalFormat = GL_RED;
+                   /* switch (dstFormat.getType()) {
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/GLBackendTexture.cpp]: why is this dead code still here? Can you remove it? Do you need it for something?

File: libraries/render-utils/src/occlusion_blend.slf [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+varying vec2 varTexcoord;
+
+uniform sampler2D blurredOcclusionTexture;
+
+void main(void) {
+    vec4 occlusionColor = texture2D(blurredOcclusionTexture, varTexcoord);
+    
+    if(occlusionColor.r > 0.8 && occlusionColor.r <= 1.0) {
+        gl_FragColor = vec4(vec3(0.0), 1.0);
+    }
```
Comment (ZappoMan) [inline libraries/render-utils/src/occlusion_blend.slf]: coding standard: brace placement

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1f58ad5d87365e6dae22b05e9d4fbdcb7e614a12 available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51e386aabdf2aaa1a719e8b61d9b0c28a0b46c5e available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 92595583ec280e81a3e9a5783a696bfce2b17063 available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

File: libraries/render-utils/src/AmbientOcclusionEffect.cpp [hunk 14]
```diff
@@ -0,0 +1,255 @@
+//
+//  AmbientOcclusionEffect.cpp
+//  interface/src/renderer
+//
+//  Created by Niraj Venkat on 7/14/13.
```
Comment (ZappoMan) [inline libraries/render-utils/src/AmbientOcclusionEffect.cpp]: this date is incorrect, you didn't create this file in 2013

File: libraries/render-utils/src/ambient_occlusion.slf [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    vec3 currentFragWorldSpace;
+    <$transformEyeToWorldDir(cam, currentFragEyeSpace, currentFragWorldSpace)$>
+    
+    vec3 cameraToPositionRay = normalize(currentFragWorldSpace - cameraPositionWorldSpace);
+    vec3 origin = cameraToPositionRay * depthVal + cameraPositionWorldSpace;
+  
+    vec3 normal = normalize(texture2D(normalTexture, varTexcoord).xyz);
+    //normal = normalize(normal * normalMatrix);
+  
+    vec3 rvec    = vec3(getRandom(varTexcoord.xy), getRandom(varTexcoord.yx), getRandom(varTexcoord.xx)) * 2.0 - 1.0;
```
Comment (ZappoMan) [inline libraries/render-utils/src/ambient_occlusion.slf]: coding standard: the coding standard explicitly prohibits doing spacing to line up code like this... just put one space before and after each of the = here and next 3 lines.

File: libraries/render-utils/src/ambient_occlusion.slf [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    vec3 bitangent  = cross(normal, tangent);
+    mat3 tbn        = mat3(tangent, bitangent, normal);
+  
+    float occlusion = 0.0;
+  
+    for (int i = 0; i < SAMPLE_COUNT; ++i) {
+        vec3 samplePos = origin + (tbn * sampleKernel[i]) * g_sample_rad;
+        vec4 offset = cam._projectionViewUntranslated * vec4(samplePos, 1.0);
+    
+        offset.xy   = (offset.xy / offset.w) * 0.5 + 0.5;
```
Comment (ZappoMan) [inline libraries/render-utils/src/ambient_occlusion.slf]: coding standard: no horizontal alignment

File: libraries/render-utils/src/ambient_occlusion.slf [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        float depth = length(samplePos - cameraPositionWorldSpace);
+    
+        float sampleDepthVal = texture2D(depthTexture, offset.xy).r;
+    
+        float rangeDelta  = abs(depthVal - sampleDepthVal);
```
Comment (ZappoMan) [inline libraries/render-utils/src/ambient_occlusion.slf]: coding standard: one space before =

File: libraries/render-utils/src/gaussian_blur.slf [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 0])*0.0044299121055113265;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 1])*0.00895781211794;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 2])*0.0215963866053;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 3])*0.0443683338718;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 4])*0.0776744219933;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 5])*0.115876621105;
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 6])*0.147308056121;
+    gl_FragColor += texture2D(occlusionTexture, varTexcoord         )*0.159576912161;
```
Comment (ZappoMan) [inline libraries/render-utils/src/gaussian_blur.slf]: coding standard: no horizontal alignment

File: libraries/render-utils/src/gaussian_blur.slf [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+//varying vec4 interpolatedNormal;
+
+varying vec2 varTexcoord;
+varying vec2 varBlurTexcoords[14];
+
+uniform sampler2D occlusionTexture;
+
+void main(void) {
+	gl_FragColor = vec4(0.0);
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 0])*0.0044299121055113265;
```
Comment (ZappoMan) [inline libraries/render-utils/src/gaussian_blur.slf]: coding standard: here through line 41 - there should be a space before and after the * operator

File: libraries/render-utils/src/gaussian_blur.slf [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+//varying vec4 interpolatedNormal;
+
+varying vec2 varTexcoord;
+varying vec2 varBlurTexcoords[14];
+
+uniform sampler2D occlusionTexture;
+
+void main(void) {
+    gl_FragColor = vec4(0.0);
+    gl_FragColor += texture2D(occlusionTexture, varBlurTexcoords[ 0])*0.0044299121055113265;
```
Comment (ZappoMan) [inline libraries/render-utils/src/gaussian_blur.slf]: coding standard: spaces before/after the * in all these lines

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1193b89918dba29ab25a0d5d14ee0d2c2b9b525f available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d2ee74f7c7c844783fd7558771d1aaf51efc22af available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

File: interface/src/Menu.h [hunk 21]
```diff
@@ -134,7 +134,7 @@ namespace MenuOption {
     const QString AddressBar = "Show Address Bar";
     const QString AlignForearmsWithWrists = "Align Forearms with Wrists";
     const QString AlternateIK = "Alternate IK";
-    const QString AmbientOcclusion = "Ambient Occlusion";
+    const QString AmbientOcclusion = "Debug Ambient Occlusion";
```
Comment (ZappoMan) [inline interface/src/Menu.h]: We should make this constant name DebugAmbientOcclusion and change the code in Menu.cpp to make sure it defaults to OFF. Ok? This is a merge blocker.

File: libraries/render-utils/src/RenderDeferredTask.cpp [hunk 22]
```diff
@@ -80,6 +81,11 @@ RenderDeferredTask::RenderDeferredTask() : Task() {
     _jobs.push_back(Job(new DrawLight::JobModel("DrawLight")));
     _jobs.push_back(Job(new RenderDeferred::JobModel("RenderDeferred")));
     _jobs.push_back(Job(new ResolveDeferred::JobModel("ResolveDeferred")));
+    _jobs.push_back(Job(new AmbientOcclusion::JobModel("AmbientOcclusion")));
```
Comment (ZappoMan) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: shouldn't we ONLY do this step if the menu item is checked? We want this to have zero effect on the render pipeline if not chosen.

File: interface/src/Menu.h (hunk 21 above)
Comment (venkatn93) [inline interface/src/Menu.h]: Thanks. Changing this now

File: libraries/render-utils/src/RenderDeferredTask.cpp (hunk 22 above)
Comment (venkatn93) [inline libraries/render-utils/src/RenderDeferredTask.cpp]: the next line says _jobs.back().setEnabled(false); and later enable it in RenderDefferedTask::run() depending on the state of the menu item

**Comment by hifi-gustavo** [comment]
Build console output for commit: 34e400ac67a04983976ec276711ec39ff3816fdf available at: https://gist.github.com/603dbde71e97837e36c7<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5376.dmg

**Pull request merged**